### PR TITLE
38481 download custom drivers each pod

### PR DESCRIPTION
**Pull request by jakefhyde**
## Problem

Node drivers are not downloaded in each rancher replica. Because the replicas are loadbalanced, when querying the `/assets` directory for `docker-machine-driver-*` files, the docker machine drivers cannot be downloaded on non-leader replicas because they are only installed on the leader node.

## Solution

Download the docker machine drivers on all replicas. This controller runs all the time, but will prevent processing if a node becomes the leader. This does not unregister the handler, but I do not want to introduce such a change into wrangler yet.
 
### Manual Testing

Manually tested with 3 replicas, confirmed that the machine provision job had no issues downloading the custom node driver. The driver is present in all replicas at `/usr/share/rancher/ui/assets`.

### Automated Testing

* Test types added/modified:
    * None, we do not currently have a third party node driver that we would be able to test with in CI, nor do we have a framework setup for being able to verify the filesystem contents of the rancher pods. 

## QA Testing Considerations

QA should test both the vendored in 3rd party node drivers (for example, nutanix/proxmox if possible). Cloned versions of those drivers will not work due to reasons found here: https://github.com/rancher/rancher/issues/37074
 
### Regressions Considerations

QA should test that built-in node drivers are unaffected.

Existing / newly added automated tests that provide evidence there are no regressions: N/A, not possible to add test

### CODE REVIEW

File: pkg/controllers/dashboardapi/nodedriver/nodedriver.go [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+	wrangler.Mgmt.NodeDriver().OnChange(ctx, "custom-node-driver-handler", onChange)
+
+	// when this pod becomes the leader, do not run the onChange code
+	wrangler.OnLeader(func(ctx context.Context) error {
+		leader.Store(true)
+		return nil
+	})
+}
+
+func onChange(_ string, obj *v3.NodeDriver) (*v3.NodeDriver, error) {
```
Comment (bfbachmann) [inline pkg/controllers/dashboardapi/nodedriver/nodedriver.go:24]: Maybe it would be pointless (I'm curious what your take is), but could it be worth writing a small unit test for this?

Comment (jakefhyde) [reply pkg/controllers/dashboardapi/nodedriver/nodedriver.go:24]: I can't really write a unit test that confirms expected behavior, but I can probably  write one that verifies we don't process invalid drivers.

**Review by jiaqiluo — CHANGES_REQUESTED** [review]
I feel that we should add a new condition, "FollowerDownloaded",  to the NodeDriver object, which gets updated in the event of successful downloading of the node driver to the followers. 
Also, we should use the `appliedChecksum`  `appliedURL` value in the onChange function. 

My suggestions are based on the existing implementation for downloading the ClusterRepo on [the leader](https://github.com/rancher/rancher/blob/69c5996434255f7ae8eb831fe4d0eb2fa23c2c12/pkg/controllers/dashboard/helm/repo.go#L44-L61) and [the followers](https://github.com/rancher/rancher/blob/69c5996434255f7ae8eb831fe4d0eb2fa23c2c12/pkg/controllers/dashboard/helm/repo.go#L63-L74).

**Comment by jakefhyde** [comment]
@jiaqiluo Can you explain what you mean by
> Also, we should use the `appliedChecksum` `appliedURL` value in the `onChange` function.

Also the reason I'm hesitant to implement a similar paradigm is that in the case of multiple replicas, having a `FollowerDownloaded` condition will cause the followers to race, such that if the first replica fails to download, it will update the object with the failure, the second replica will have succeeded the download, but fail to update the status because it was previously updated by the failing replica, causing an unnecessary re-reconciliation, as well as consuming the potential error the next time the object is enqueued on the second replica. I feel strongly that we shouldn't have multiple replicas racing to update a field on an object. WDYT?

**Comment by jiaqiluo** [comment]
@jakefhyde, 

> Can you explain what you mean by
>    Also, we should use the appliedChecksum appliedURL value in the onChange function.

This comment is only valid if we go with the updating-status approach.  And I mean we should use the value from `.status.appliedChecksum` and `.status.appliedURL` on the NodeDriver object when downloading the node driver on the followers. You can ignore it with your current implication. 

I understand your points about the potential conflicts. I think the existing generated status handler addresses it: because in both your case and the cluster repo case, we do not update the metadata or spec part of the object, the change of a conflicting error is rare; even if a conflicting error happens, the status handler can still [update the target status to reflect the changes](https://github.com/rancher/rancher/blob/69c5996434255f7ae8eb831fe4d0eb2fa23c2c12/pkg/generated/controllers/catalog.cattle.io/v1/clusterrepo.go#L98-L122). So I think a few more reconciliation is worthy when considering the benefit that the new condition can reflect the actual status of the downloading in followers. 


If we will go with your current implementation, I have the following two cents (hilariously there are literally two): 
1. There is no indication of whether the followers' downloading is finished successfully or not. The nodeDriver's `spec.Active` field reflects whether the downloading on the leader finishes.  There is still a gap, and if users create a new cluster by using the node driver and the request is load balanced to the follower which has not finished the downloading, the user will hit the exact issue that the PR tries to fix. This might not happen often, but it is still possible. 
3. The current implementation will skip the followers' downloading if the value of `obj.Spec.Active`  is true, which means if the 1st attempt of downloading on the follower fails but the same downloading success on the leader, the follower will not have a chance to download the node driver again because `obj.Spec.Active`  has been true.

**Comment by jakefhyde** [comment]
I think you should check the logic again there, I'm only downloading the driver onto follower replicas if the node driver is active.
```
	if !obj.Spec.Active {
		return obj, nil
	}
```
Active doesn't quite mean that the node driver is downloaded, it is a requirement for a node driver to be downloaded, but they don't mean the same thing. The existing code isn't very intuitive, but the following takes care of only downloading if the hash has changed:
```
	if err := driver.Stage(true); err != nil {
		return nil, err
	}

	if err := driver.Install(); err != nil {
		return nil, err
	}

	if err := driver.Executable(); err != nil {
		return nil, err
	}
```

Also, there is definitely an edge case with downloading the driver, however the probability is very low due to the fact that you have to be provisioning while new replicas are spinning up. For what it's worth, attempting to use conditions also doesn't solve that problem, because you are relying on the fact that multiple replicas aren't thrashing and that the condition is the most up to date, which isn't always guaranteed, i.e. there is no association between which followers have downloaded the objects. A secondary fix will likely be going into the rancher-machine image to add a configurable (default 5) number of retries to download the asset, which will make this edge case impossible.

File: pkg/controllers/dashboardapi/nodedriver/nodedriver.go [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+	if obj.Spec.Builtin {
+		return obj, nil
+	}
+
+	driver := drivers.NewDynamicDriver(obj.Spec.Builtin, obj.Spec.DisplayName, obj.Spec.URL, obj.Spec.Checksum)
+	if driver.Exists() {
+		return obj, nil
+	}
+
+	if err := driver.Stage(true); err != nil {
```
Comment (jiaqiluo) [inline pkg/controllers/dashboardapi/nodedriver/nodedriver.go:48]: Question:  Will the "true" here cause unnecessary redownloading of the same driver?  Should we do some checks like [what this function](https://github.com/rancher/rancher/blob/8e8c62c86d53690a77260884e6fe1afcdafaafc4/pkg/controllers/management/drivers/nodedriver/machine_driver.go#L298-L324) does?

**Review by jiaqiluo — CHANGES_REQUESTED** [review]
One last question

File: pkg/controllers/dashboardapi/nodedriver/nodedriver.go (hunk 2 above)
Comment (jakefhyde) [reply pkg/controllers/dashboardapi/nodedriver/nodedriver.go:48]: This code is what makes it unnecessary: https://github.com/rancher/rancher/blob/c294a41c4253dae7d8de4dfb4ba556134e97b728/pkg/controllers/management/drivers/base_driver.go#L104-L110. The `cacheFile` is in `management-state/machine-drivers/<sha25sum of the URL + driver hash>`, which if it exists means the file is not redownloaded. Now that you mention it, it may be worth gating the `Install` and `Executable` functions behind this `cacheFile` as well, since they don't have similar checks and copy the downloaded binary to the expected path.

**Review by jiaqiluo — APPROVED** [review]
LGTM

File: pkg/controllers/dashboardapi/nodedriver/nodedriver.go (hunk 2 above)
Comment (jiaqiluo) [reply pkg/controllers/dashboardapi/nodedriver/nodedriver.go:48]: OK, it sounds like a good enhancement to do.

File: pkg/rancher/rancher.go [hunk 3]
```diff
@@ -261,6 +262,10 @@ func (r *Rancher) Start(ctx context.Context) error {
 		return err
 	}
 
+	if features.MCM.Enabled() {
```
Comment (jakefhyde) [inline pkg/rancher/rancher.go:265]: Changed registration from `dashboardapi` to top level, as the nodedriver handler does not need to run in the clsuter agent (and was breaking CI).

File: scripts/test [hunk 4]
```diff
@@ -80,10 +80,7 @@ run_rancher()
     while sleep 2; do
```
Comment (jakefhyde) [inline scripts/test:80]: Added the output logs akin to the provisioning tests. Maybe we combine the logic there someday, but for now this should suffice.

File: pkg/controllers/management/drivers/nodedriver/machine_driver.go [hunk 5]
```diff
@@ -84,24 +84,27 @@ type Lifecycle struct {
...
-func (m *Lifecycle) download(obj *v3.NodeDriver) (*v3.NodeDriver, error) {
+func (m *Lifecycle) download(obj *v32.NodeDriver) (*v32.NodeDriver, error) {
 	driverLock.Lock()
 	defer driverLock.Unlock()
 	if !obj.Spec.Active && !obj.Spec.AddCloudCredential {
 		return obj, nil
 	}
 
 	forceUpdate := m.checkDriverVersion(obj)
+	if v32.NodeDriverConditionDownloaded.GetStatus(obj) == "" || v32.NodeDriverConditionInstalled.GetStatus(obj) == "" {
```
Comment (jakefhyde) [inline pkg/controllers/management/drivers/nodedriver/machine_driver.go:99]: I added this in the event the driver was downloaded, but the status wasn't updated, which is really to deal with edge cases.

**Review by bfbachmann — APPROVED** [review]
LGTM! Thanks for adding the test

**Review by bfbachmann — APPROVED** [review]

**Review by jiaqiluo — APPROVED** [review]
